src/bases/createWidgetBase.ts

```diff
@@ -0,0 +1,185 @@
+import { ComposeFactory } from 'dojo-compose/compose';
+import createStateful from 'dojo-compose/bases/createStateful';
+import {
+	HNode,
+	DNode,
+	WNode,
+	Widget,
+	WidgetMixin,
+	WidgetState,
+	WidgetOptions
+} from 'dojo-interfaces/widgetBases';
+import { VNode, VNodeProperties } from 'dojo-interfaces/vdom';
+import { Factory } from 'dojo-interfaces/core';
+import { assign } from 'dojo-core/lang';
+import WeakMap from 'dojo-shim/WeakMap';
+import d from './../util/d';
+
+export interface WidgetFactory extends ComposeFactory<Widget<WidgetState>, WidgetOptions<WidgetState>> {}
+
+interface WidgetInternalState {
+	readonly id?: string;
+	dirty: boolean;
+	widgetClasses: string[];
+	cachedWidget?: VNode;
+	historicChildrenMap: Map<string | Factory<Widget<WidgetState>, WidgetOptions<WidgetState>>, Widget<WidgetState>>;
+	currentChildrenMap: Map<string | Factory<Widget<WidgetState>, WidgetOptions<WidgetState>>, Widget<WidgetState>>;
+};
+
+const widgetInternalStateMap = new WeakMap<Widget<WidgetState>, WidgetInternalState>();
+
+function isWNode(child: DNode): child is WNode {
+	return child && (<WNode> child).factory !== undefined;
+}
+
+function realizeDNode(instance: Widget<WidgetState>, dNode: DNode) {
```

Comment:
I think the return annotation would make it clearer.  It is required for exported functions, but unless it is bloody obvious what it does then even internal functions should be explicit (it also help safeguard the function returns from logic errors, especially in strict mode).  It does feel a bit complex to maintain, so yes, 👍 on splitting it.  I am not a fan of the term `realize` (even though we use it in `app`)... this actually feels like to me 🚲 🏠  `resolveVNode()` or `dNodeToVNode()` or `getVNode()`.